### PR TITLE
ENTITY-2640 - Keycloak/SSO Integration

### PR DESCRIPTION
**Pull request by bsnopek-freshworks**
*Issue #:* /bcgov/entity#2640

*Description of changes:*

- Added the SBC-COMMON-COMPONENT for signin/signout
- Updated the TS version to allow for optional chaining (as required by sbc-common)
- Added enumerators for session-storage, roles, pages, as required by sbc-common
- Added ENV props (some now duplicated) as required by sbc-common
- Added routes, and routing middleware to handle requiresAuth routes

NOTE:
- temporarily set the defaulty route to signin until a non-auth route is available (need designs)
- no unit-tests yet - should get the above first

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the bcrs-entities-create-ui license (Apache 2.0).

### CODE REVIEW

File: src/enums/authPages.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+export enum Pages {
```
Comment (leksmall) [inline src/enums/authPages.ts]: Do we need this enum? Looks like auth specific pages

File: src/views/auth/Signin.vue [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    SbcSignin
+  }
+})
+export default class SigninView extends Vue {
+  private async mounted () {
+    console.log('[SigninView] -- Mounted')
+  }
+  private async redirectOnLogin () {
+    console.log('[SigninView] -- kebab-style-found')
+    this.$router.push('/draft-company')
```
Comment (leksmall) [inline src/views/auth/Signin.vue]: Is this a valid route

File: package.json [hunk 3]
```diff
@@ -14,10 +14,9 @@
     "axios": "^0.18.1",
     "core-js": "^3.6.4",
     "css-loader": "^0.28.11",
-    "launchdarkly-js-client-sdk": "^2.17.0",
     "lodash": "^4.17.15",
     "regenerator-runtime": "^0.13.3",
-    "sbc-common-components": "2.0.41",
+    "sbc-common-components": "^2.1.8",
```
Comment (severinbeauvais) [inline package.json]: Don't use ^ ... we don't want to automatically upgrade to the latest version. We want to control when we upgrade as there is extra regression testing to do.

File: package.json [hunk 4]
```diff
@@ -43,7 +42,7 @@
     "eslint-plugin-vue": "^5.2.3",
     "sass": "^1.25.0",
     "sass-loader": "^8.0.0",
-    "typescript": "<3.6.0",
+    "typescript": "3.7.3",
```
Comment (severinbeauvais) [inline package.json]: We didn't want to go past 3.6.0 as that broke other package dependencies. I note that you aren't updating other packages so something is probably broken somewhere or at least you're getting npm or build warnings/errors.

There's another ticket to upgrade TS because it's not as straightforward as this unfortunately.

I will try to help with the packages on Monday.

File: src/App.vue [hunk 5]
```diff
@@ -147,6 +147,11 @@ export default class App extends Mixins(DateMixin, FilingTemplateMixin, LegalApi
       .find(x => x.entityType === val) : null)
   }
 
+  // Bound to the successful login event in SigninView
+  private redirectOnLogin (): void {
+    alert('in app')
```
Comment (severinbeauvais) [inline src/App.vue]: Ok for debugging but don't commit this to master branch.

File: src/App.vue [hunk 6]
```diff
@@ -155,7 +160,4 @@ export default class App extends Mixins(DateMixin, FilingTemplateMixin, LegalApi
 </script>
 
 <style lang="scss" scoped>
-.app-header {
```
Comment (severinbeauvais) [inline src/App.vue]: Why remove this?

File: src/assets/styles/layout.scss [hunk 7]
```diff
@@ -33,7 +33,7 @@
 }
 
 .container {
-  max-width: 1360px; // should match auth-web, etc
```
Comment (severinbeauvais) [inline src/assets/styles/layout.scss]: This was a recent fix that you're undoing. Please revert this change.

File: src/components/common/Stepper.vue [hunk 8]
```diff
@@ -1,17 +1,12 @@
 <template>
-  <v-container id="step-buttons-container">
```
Comment (severinbeauvais) [inline src/components/common/Stepper.vue]: Please revert your changes to this file.

File: src/App.vue [hunk 9]
```diff
@@ -155,7 +160,4 @@ export default class App extends Mixins(DateMixin, FilingTemplateMixin, LegalApi
...
-  z-index: 1000;
```
Comment (severinbeauvais) [inline src/App.vue]: This was a recent change that you're undoing. Please revert this change.

File: src/App.vue [hunk 10]
```diff
@@ -96,24 +96,24 @@ export default class App extends Mixins(DateMixin, FilingTemplateMixin, LegalApi
...
-      // Parse the draft data into the store if it exists
-      this.draftFiling && this.parseDraft(this.draftFiling)
-
-      // Inform the router view we are resuming a draft and to update ui
-      this.isDraft = true
-    } catch (e) {
-      // TODO: Catch a flag from the api, if there is an error to be handled.
-    }
-  }
+  // private async created (): Promise<any> {
```
Comment (severinbeauvais) [inline src/App.vue]: Ok for debugging but please work with Cameron on a solution.

File: package.json [hunk 11]
```diff
@@ -14,10 +14,9 @@
     "axios": "^0.18.1",
     "core-js": "^3.6.4",
     "css-loader": "^0.28.11",
-    "launchdarkly-js-client-sdk": "^2.17.0",
```
Comment (severinbeauvais) [inline package.json]: This was a recent fix that you're undoing. Please revert this change.

File: src/main.ts [hunk 12]
```diff
@@ -10,7 +10,6 @@ import router from '@/router'
 import { store } from '@/store'
 import Affix from 'vue-affix'
 import Vue2Filters from 'vue2-filters' // needed by SbcFeeSummary
-import { featureFlags, initLDClient } from '@/common/FeatureFlags'
```
Comment (severinbeauvais) [inline src/main.ts]: This was a recent fix that you're undoing. Please revert this change.

File: src/main.ts [hunk 13]
```diff
@@ -63,22 +64,21 @@ fetchConfig()
     // const authzSerice = new AuthorizationService() // FUTURE
     // await authzSerice.getAuthorizations('NR123456') // FUTURE
     */
-    if (featureFlags.getFlag('bcrs-create-ui-enabled')) {
```
Comment (severinbeauvais) [inline src/main.ts]: This was a recent fix that you're undoing. Please revert this change.

File: src/main.ts [hunk 14]
```diff
@@ -63,22 +64,21 @@ fetchConfig()
...
+    }).$mount('#app')
   })
   .catch((error: string) => {
     /**
      * This catches any un-handled errors from fetchConfig()
      * or anything else in then() block above.
      */
     console.error(error) // eslint-disable-line no-console
-    alert('There was an error starting this page. (See console for details.)' +
+    alert('There was an errors starting this page. (See console for details.)' +
```
Comment (severinbeauvais) [inline src/main.ts]: Typo.

File: public/config/configuration.json [hunk 15]
```diff
@@ -6,5 +6,8 @@
   "PAY_API_URL": "https://pay-api-dev.pathfinder.gov.bc.ca/api/v1/",
   "KEYCLOAK_CONFIG_URL": "/local-keycloak-config-url/keycloak.json",
   "ADDRESS_COMPLETE_KEY": "RH99-RN99-GB72-FW86",
-  "LD_CLIENT_ID": "5e5d5af4574f7e09ab53d03b"
+  "VUE_APP_PAY_ROOT_API": "https://pay-api-dev.pathfinder.gov.bc.ca/api/v1",
```
Comment (severinbeauvais) [inline public/config/configuration.json]: I think it's ok to rename these if that's the intent, but I don't think it's ok to have duplicate keys (tech debt). Please clean up one way or the other.

File: src/router/index.ts [hunk 16]
```diff
@@ -14,4 +18,37 @@ const router = new VueRouter({
...
 })
 
+router.beforeEach((to, from, next) => {
+  // If the route requires authentication
+  if (to.matched.some(record => record.meta.requiresAuth)) {
+    // If we have a token
+    if (sessionStorage.getItem(SessionStorageKeys.KeyCloakToken)) {
+      // If there are allowed or disabled roles specified on the route, check if
+      //  the user has those roles, else route to signout
+      if (!KeyCloakService.verifyRoles(to.meta.allowedRoles, to.meta.disabledRoles)) {
```
Comment (severinbeauvais) [inline src/router/index.ts]: We don't specify `allowedRoles` or `disabledRoles` on our routes. Do we need to? If not then this code can be greatly simplified.

File: src/utils/config-helper.ts [hunk 17]
```diff
@@ -55,10 +57,6 @@ export function fetchConfig (): Promise<any> {
       const addressCompleteKey = response.data['ADDRESS_COMPLETE_KEY'];
       (<any>window).addressCompleteKey = addressCompleteKey
       console.info('Set Address Complete Key.')
-
```
Comment (severinbeauvais) [inline src/utils/config-helper.ts]: This was a recent fix that you're undoing. Please revert this change.

File: src/utils/config-helper.ts [hunk 18]
```diff
@@ -44,6 +44,8 @@ export function fetchConfig (): Promise<any> {
       sessionStorage.setItem('AUTH_API_URL', authApiUrl)
       console.log('Set Auth API URL to: ' + authApiUrl)
 
+      sessionStorage.setItem('VUE_APP_AUTH_ROOT_API', response.data['VUE_APP_AUTH_ROOT_API'])
```
Comment (severinbeauvais) [inline src/utils/config-helper.ts]: I think you can use `AUTH_API_URL` instead.

File: src/views/auth/Signout.vue [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+// Components
+import SbcSignout from 'sbc-common-components/src/components/SbcSignout.vue'
+
+@Component({
+  components: {
+    SbcSignout
+  }
+})
+export default class SignoutView extends Vue {
+  private redirectUrl = `${window.location.origin}/${process.env.VUE_APP_PATH}`
```
Comment (severinbeauvais) [inline src/views/auth/Signout.vue]: I suggest you redirect to `AUTH_URL` instead. (Specifically a configured value instead of an assumed calculated URL.)

**Comment by severinbeauvais** [comment]
> Is the launch darkly feature flag check/code supposed to be removed? Not totally in the loop on what is happening with launch darkly.
> 
> @leksmall @severinbeauvais @cameron-freshworks

No.

File: public/config/configuration.json [hunk 20]
```diff
@@ -6,5 +6,8 @@
   "PAY_API_URL": "https://pay-api-dev.pathfinder.gov.bc.ca/api/v1/",
   "KEYCLOAK_CONFIG_URL": "/local-keycloak-config-url/keycloak.json",
   "ADDRESS_COMPLETE_KEY": "RH99-RN99-GB72-FW86",
-  "LD_CLIENT_ID": "5e5d5af4574f7e09ab53d03b"
```
Comment (severinbeauvais) [inline public/config/configuration.json]: Please restore this key that you removed.

File: src/enums/sessionStorageKeys.ts [hunk 21]
```diff
@@ -0,0 +1,12 @@
+export enum SessionStorageKeys {
```
Comment (severinbeauvais) [inline src/enums/sessionStorageKeys.ts]: We shouldn't need this. The shared components should make use of this code in their repo, not ours.

File: src/views/auth/Signout.vue (hunk 19 above)
Comment (severinbeauvais) [reply src/views/auth/Signout.vue]: On signout, the user should be sent back to the Business Registries home page, which will be bcregistry.gov.bc.ca/business but is currently https://dev.bcregistry.ca/cooperatives/auth/.

File: src/enums/authPages.ts (hunk 1 above)
Comment (bsnopek-freshworks) [reply src/enums/authPages.ts]: This one can be removed, yes. I added it in hopes that it might have addressed some errors coming from the common-components, but it did not, so can be removed !

File: public/config/configuration.json (hunk 20 above)
Comment (bsnopek-freshworks) [reply public/config/configuration.json]: This one actually is the original key - git is picking it up because of the comma at the end which was added. ;)

File: src/enums/sessionStorageKeys.ts (hunk 21 above)
Comment (bsnopek-freshworks) [reply src/enums/sessionStorageKeys.ts]: I debated about adding this one, as resolving the magic-string antipattern for property-keys can be a bit of a russian-nesting-doll issue - but its the keys that need to be specifically set here, and this resolves that antipattern with this usage - I figured if we were to implement a means of addressing it - it might as well be the same format as used with the common-components.

I could remove it without having that significant of an impact on our codebase though, as we're only interacting with these specific keys in limited locations - however, generally if its repeated more than once, and either exposes or consumes an external dependancy - thats generally the use-case for when we'd want to implement this type of fix, no ?

File: src/views/auth/Signin.vue (hunk 2 above)
Comment (bsnopek-freshworks) [reply src/views/auth/Signin.vue]: Corrected in the latest commit, thanks !

File: src/App.vue (hunk 5 above)
Comment (bsnopek-freshworks) [reply src/App.vue]: removed in latest commit

File: src/assets/styles/layout.scss (hunk 7 above)
Comment (bsnopek-freshworks) [reply src/assets/styles/layout.scss]: poor rebase on my part, corrected in latest commit

File: src/components/common/Stepper.vue (hunk 8 above)
Comment (bsnopek-freshworks) [reply src/components/common/Stepper.vue]: poor rebase on my part, corrected in latest commit

File: src/App.vue (hunk 9 above)
Comment (bsnopek-freshworks) [reply src/App.vue]: poor rebase on my part, corrected in latest commit

File: package.json (hunk 11 above)
Comment (bsnopek-freshworks) [reply package.json]: poor rebase on my part, corrected in latest commit

File: src/main.ts (hunk 12 above)
Comment (bsnopek-freshworks) [reply src/main.ts]: poor rebase on my part, corrected in latest commit

File: src/main.ts (hunk 13 above)
Comment (bsnopek-freshworks) [reply src/main.ts]: poor rebase on my part, corrected in latest commit

File: src/main.ts (hunk 14 above)
Comment (bsnopek-freshworks) [reply src/main.ts]: poor rebase on my part, corrected in latest commit

File: public/config/configuration.json (hunk 15 above)
Comment (bsnopek-freshworks) [reply public/config/configuration.json]: Reduced to the single property, will look at merging/renaming as a dept-ticket in the (very) near future

File: src/router/index.ts (hunk 16 above)
Comment (bsnopek-freshworks) [reply src/router/index.ts]: Removed - token validation is done through the api, front-end simply checks for its existence.

File: src/utils/config-helper.ts (hunk 17 above)
Comment (bsnopek-freshworks) [reply src/utils/config-helper.ts]: poor rebase on my part, corrected in latest commit

File: src/utils/config-helper.ts (hunk 18 above)
Comment (bsnopek-freshworks) [reply src/utils/config-helper.ts]: Same as above - we'll look at renaming as a part of a tech-debt ticket in the near future

File: src/views/auth/Signout.vue (hunk 19 above)
Comment (bsnopek-freshworks) [reply src/views/auth/Signout.vue]: As per our discussion, we'll look at correcting this in a (near) future ticket.

File: src/App.vue (hunk 10 above)
Comment (bsnopek-freshworks) [reply src/App.vue]: This one still requires some work - I'll get with Cameron though and see what we can do to resolve it properly

File: src/router/index.ts [hunk 22]
```diff
@@ -14,4 +18,19 @@ const router = new VueRouter({
   }
 })
 
+router.beforeEach((to, from, next) => {
+  // If the route requires authentication
+  if (to.matched.some(record => record.meta.requiresAuth)) {
+    // If we have a token dont have a token
```
Comment (bsnopek-freshworks) [inline src/router/index.ts]: This comment needs correction ..

File: src/App.vue (hunk 10 above)
Comment (severinbeauvais) [reply src/App.vue]: Commit with this code commented out, and created a new ticket to resolve this separately.

File: src/router/index.ts (hunk 22 above)
Comment (severinbeauvais) [reply src/router/index.ts]: Please add "FUTURE" comment indicating that we should handle token expiry at some point.

File: src/enums/sessionStorageKeys.ts (hunk 21 above)
Comment (severinbeauvais) [reply src/enums/sessionStorageKeys.ts]: Reach into sbc-common-components package to ref this enum where you need it.

File: src/enums/authRoles.ts [hunk 23]
```diff
@@ -0,0 +1,6 @@
+export enum Role {
```
Comment (severinbeauvais) [inline src/enums/authRoles.ts]: Not needed?

File: src/App.vue (hunk 10 above)
Comment (bsnopek-freshworks) [reply src/App.vue]: Created entity #2958 for this task

File: package.json (hunk 4 above)
Comment (severinbeauvais) [reply package.json]: If you rebase to the latest master, the Typescript package (and many others) have been updated, so you should not have to change this line.

File: src/utils/config-helper.ts (hunk 18 above)
Comment (bsnopek-freshworks) [reply src/utils/config-helper.ts]: A new ticket has been created for this - Entity #2960

File: src/enums/authRoles.ts (hunk 23 above)
Comment (bsnopek-freshworks) [reply src/enums/authRoles.ts]: Removed in latest commit

File: src/enums/sessionStorageKeys.ts (hunk 21 above)
Comment (bsnopek-freshworks) [reply src/enums/sessionStorageKeys.ts]: Resolved in latest commit

File: src/router/index.ts (hunk 22 above)
Comment (bsnopek-freshworks) [reply src/router/index.ts]: Resolved in latest commit

File: src/router/routes.ts [hunk 24]
```diff
@@ -3,36 +3,59 @@
...
+    path: '/signout',
+    name: 'signout',
+    component: SignoutView,
+    props: true,
+    meta: {
+      requiresAuth: true
+    }
+  },
+  {
+    path: '/signout/:redirectUrl',
```
Comment (severinbeauvais) [inline src/router/routes.ts]: 1. How do we get to this route (page)?

2. Does this page require auth (see line 56)?

Comment (bsnopek-freshworks) [reply src/router/routes.ts]: This is used by the common-component in the header - there is a signout option there and it expects this route.  It requires authentication because it intends to remove the tokens from sessionStorage - however - is the common-component smart enough to handle a missing token gracefully ? Does it need to be a valid token ? These I do not know..

File: src/router/routes.ts [hunk 25]
```diff
@@ -3,36 +3,59 @@
...
+    path: '/signin/:idpHint',
+    name: 'signin',
+    component: SigninView,
+    props: true,
+    meta: {
+      requiresAuth: false
+    }
+  },
+  {
+    path: '/signin/:idpHint/:redirectUrl',
```
Comment (severinbeauvais) [inline src/router/routes.ts]: Is it possible in the PR to remove the other 2 signin routes? (Per my RC text where I noted that only the second route is used -- the one with 2 route parameters.)

File: src/router/routes.ts [hunk 26]
```diff
@@ -3,36 +3,59 @@
...
+    path: '/signin/:idpHint/:redirectUrl/:redirectUrlLoginFail',
+    name: 'signin-redirect-fail',
+    component: SigninView,
+    props: true,
+    meta: {
+      requiresAuth: false
+    }
+  },
+  {
+    path: '/signout',
```
Comment (severinbeauvais) [inline src/router/routes.ts]: How do we get to this route? From the header component? May need to check with Andrew for this.

File: src/views/auth/Signout.vue (hunk 19 above)
Comment (severinbeauvais) [reply src/views/auth/Signout.vue]: No, don't send user back to our app (which will then redirect to the signin page).

Instead, send them "up" to the Business Registries home page (per comment above).

File: src/router/routes.ts (hunk 26 above)
Comment (severinbeauvais) [inline src/router/routes.ts]: 1. How do we get to this signout route (page)?

2. Does this page require auth (see line 56)? ie, do we need to be logged in before we can go to the signout page?

File: src/enums/authPages.ts (hunk 1 above)
Comment (severinbeauvais) [reply src/enums/authPages.ts]: Has been removed.

File: src/router/routes.ts (hunk 26 above)
Comment (severinbeauvais) [reply src/router/routes.ts]: This route is set by the SbcHeader component ... but please add a code comment so this makes sense to the next developer.

File: src/router/index.ts [hunk 27]
```diff
@@ -14,4 +18,21 @@ const router = new VueRouter({
...
 
+router.beforeEach((to, from, next) => {
+  // If the route requires authentication
+  if (to.matched.some(record => record.meta.requiresAuth)) {
+    // If we  dont have a token
+    if (!sessionStorage.getItem(SessionStorageKeys.KeyCloakToken)) {
+      // Send them to the bcsc login
+      return next({
+        path: `/signin/bcsc${to.path}`,
+        query: { redirect: to.fullPath }
```
Comment (severinbeauvais) [inline src/router/index.ts]: If this line of code has the effect of passing the current route (eg, define-company) to the signin component, then the signin component can route back to it after successful login.

Comment (severinbeauvais) [reply src/router/index.ts]: (So that the signin component is no longer hard-coded to redirect to the define-business route.)

Comment (severinbeauvais) [reply src/router/index.ts]: There's some question how this works and whether my proposed change will handle all use cases, so let's go with the existing code for now and change later if needed.

File: src/router/routes.ts (hunk 26 above)
Comment (severinbeauvais) [reply src/router/routes.ts]: I think we probably don't need to specify that the signout page requires auth, but it doesn't break anything either way, so this is OK for now and we can change it later if needed.

Comment (severinbeauvais) [reply src/router/routes.ts]: Someone can add such a comment in a future commit (since this code is subject to change anyway with the NR query and draft resuming that are also being implemented at the moment).

Comment (severinbeauvais) [reply src/router/routes.ts]: Done.

**Comment by severinbeauvais** [comment]
> Conditional approval -- there are a couple of outstanding comments:
> 
> 1. I'd like to see a comment on the signout route to let future developers know how this route is used (ie, how we get to this route)
> 2. I'd like the signin component to not be hard-coded to redirect to the define-company route, which I believe can be done with most of the existing code in the router and in the signin page

Approved.

**Comment by cameron-freshworks** [comment]
3 approvals and no outstanding issues. Merging!
